tensorflow/lite/interpreter_builder.cc

```diff
@@ -480,6 +481,20 @@ TfLiteStatus InterpreterBuilder::ParseTensors(
           if (size_t size = array->size()) {
             *buffer_size = size;
             *buffer_data = reinterpret_cast<const char*>(array->data());
+            if (need_to_swap_bytes_) {
+              switch (type) {
+                case kTfLiteFloat32: {
+                  int32_t *p = reinterpret_cast<int32_t*>(const_cast<uint8_t*>(array->data()));
```

Comment:
I don't think this is safe, as the buffer here could be from a mmap'ed file? I think we'd need to swizzle all the constant buffers into new staging buffers.

Reply:
Sure, that makes sense. We will push the changes along with another approach we are testing right now, thanks for the suggestion.

Reply:
Sounds good, thanks! I'm happy to help sketch out a design on how this could work, or if you want to put together a one-pager proposal where we can discuss, that might be better than a pull request?

Reply:
Thanks @jdduke for offering to sketch out a design on how this might work. We are still learning the internals of how it fits together so any help is appreciated. My thought is to check [https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/framework/graph_util_impl.py#L449](url) for system endianness and byte-swapping the tensor data for non-LE endiness during reconstruction of graph. This way the model.tflite file will behave as created on LE system irrespective of the actual endianness of the system. I can then byte-swap tensor during parsing of tensors in [https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/lite/model.cc#L596](url) guarding it with FLATBUFFER_LITTLEENDIAN flag.

Reply:
Thanks @jdduke for offering to sketch out a design on how this might work. We are still learning the internals of how it fits together so any help is appreciated. My thought is to check [graph_util_impl.py](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/framework/graph_util_impl.py) for system endianness and byte-swapping the tensor data for non-LE endiness during reconstruction of graph. This way the model.tflite file will behave as created on LE system irrespective of the actual endianness of the system. I can then byte-swap tensor during parsing of tensors in [model.cc](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/lite/model.cc) guarding it with FLATBUFFER_LITTLEENDIAN flag.

Reply:
Yeah, we'll probably have to audit all the paths from TF data to TFLite flatbuffer data to ensure LE compat. Looping in @abattery and @karimnosseir as FYI.

> I can then byte-swap tensor during parsing of tensors in model.cc guarding it with FLATBUFFER_LITTLEENDIAN flag.

Yeah, the main thing we'll need for that is to allocate new storage for the byte-swapped data. We've considered doing that before in the event that we encounter a model that's not properly aligned, and this is a variant of that.

Reply:
Thanks @jdduke. I did a quick prototype without creating new storage and it seem to have worked for the failing testcase.  Here are my changes:
`diff --git a/tensorflow/python/framework/convert_to_constants.py b/tensorflow/python/framework/convert_to_constants.py
index 10541ed8e3..85aa8a277e 100644
--- a/tensorflow/python/framework/convert_to_constants.py
+++ b/tensorflow/python/framework/convert_to_constants.py
@@ -871,11 +871,16 @@ class _SessionConverterData(_ConverterData):
         nodes_to_convert.append(node)
         tensor_names_to_convert.append(tensor_name + ":0")

+    print("Simrit Inside _SessionConverterData")
     if tensor_names_to_convert:
       converted_tensors = session.run(tensor_names_to_convert)
       for node, tensor_value in zip(nodes_to_convert, converted_tensors):
-        self._tensor_data[node.name] = _TensorData(
-            numpy=tensor_value, dtype=node.attr["dtype"].type, index=None)
+        if sys.byteorder == "little":
+            self._tensor_data[node.name] = _TensorData(
+                numpy=tensor_value, dtype=node.attr["dtype"].type, index=None)
+        else:
+            self._tensor_data[node.name] = _TensorData(
+                numpy=np.float32.byteswap(tensor_value), dtype=node.attr["dtype"].type, index=None)


 def disable_lower_using_switch_merge(graph_def):
diff --git a/tensorflow/lite/interpreter_builder.cc b/tensorflow/lite/interpreter_builder.cc
index f5c8d97b96..8e2437ae1f 100644
--- a/tensorflow/lite/interpreter_builder.cc
+++ b/tensorflow/lite/interpreter_builder.cc
@@ -480,6 +479,43 @@ TfLiteStatus InterpreterBuilder::ParseTensors(
           if (size_t size = array->size()) {
             *buffer_size = size;
             *buffer_data = reinterpret_cast<const char*>(array->data());
+           if(!FLATBUFFERS_LITTLEENDIAN) {
+            switch (type) {
+                  case kTfLiteFloat16:
+                  case kTfLiteInt16: {
+                       int16_t *p = reinterpret_cast<int16_t*>(const_cast<uint8_t*>(array->data()));
+                       for (size_t i = 0; i < size; i+=2, ++p)
+                             *p = flatbuffers::EndianScalar(*p);
+                       break;
+                       }
+                  case kTfLiteInt32:
+                  case kTfLiteFloat32: {
+                       int32_t *p = reinterpret_cast<int32_t*>(const_cast<uint8_t*>(array->data()));
+                       for (size_t i = 0; i < size; i+=4, ++p)
+                             *p = flatbuffers::EndianScalar(*p);
+                       break;
+                       }
+                  case kTfLiteInt64:
+                  case kTfLiteComplex64: {
+                       int64_t *p = reinterpret_cast<int64_t*>(const_cast<uint8_t*>(array->data()));
+                       for (size_t i = 0; i < size; i+=8, ++p)
+                             *p = flatbuffers::EndianScalar(*p);
+                       break;
+                       }
+                  case kTfLiteNoType:
+                  case kTfLiteUInt8:
+                  case kTfLiteString:
+                  case kTfLiteBool:
+                  case kTfLiteInt8: {
+                        break;
+                        }
+                  default: {
+                        error_reporter_->Report(
+                                "Tensor type %d incorrect.\n", type);
+                        return kTfLiteError;
+                        }
+                  }
+            }
             return kTfLiteOk;
           }
         }

`
However, I would be more comfortable if this part is handled by SME like yourself so that TF data paths are appropriately covered. If you think it is a reasonable request then I will be happy to assist with testing the changes on BE platform.
Let me know your thoughts. Thanks.

Reply:
Thanks @jdduke. I did a quick prototype without creating new storage and it seem to have worked for the failing testcase.  Here are my changes:
```
diff --git a/tensorflow/lite/interpreter_builder.cc b/tensorflow/lite/interpreter_builder.cc
index f5c8d97b96..8e2c5fceb4 100644
--- a/tensorflow/lite/interpreter_builder.cc
+++ b/tensorflow/lite/interpreter_builder.cc
@@ -480,6 +480,43 @@ TfLiteStatus InterpreterBuilder::ParseTensors(
           if (size_t size = array->size()) {
             *buffer_size = size;
             *buffer_data = reinterpret_cast<const char*>(array->data());
+           if(!FLATBUFFERS_LITTLEENDIAN) {
+            switch (type) {
+                  case kTfLiteFloat16:
+                  case kTfLiteInt16: {
+                       int16_t *p = reinterpret_cast<int16_t*>(const_cast<uint8_t*>(array->data()));
+                       for (size_t i = 0; i < size; i+=2, ++p)
+                             *p = flatbuffers::EndianScalar(*p);
+                       break;
+                       }
+                  case kTfLiteInt32:
+                  case kTfLiteFloat32: {
+                       int32_t *p = reinterpret_cast<int32_t*>(const_cast<uint8_t*>(array->data()));
+                       for (size_t i = 0; i < size; i+=4, ++p)
+                             *p = flatbuffers::EndianScalar(*p);
+                       break;
+                       }
+                  case kTfLiteInt64:
+                  case kTfLiteComplex64: {
+                       int64_t *p = reinterpret_cast<int64_t*>(const_cast<uint8_t*>(array->data()));
+                       for (size_t i = 0; i < size; i+=8, ++p)
+                             *p = flatbuffers::EndianScalar(*p);
+                       break;
+                       }
+                  case kTfLiteNoType:
+                  case kTfLiteUInt8:
+                  case kTfLiteString:
+                  case kTfLiteBool:
+                  case kTfLiteInt8: {
+                        break;
+                        }
+                  default: {
+                        error_reporter_->Report(
+                                "Tensor type %d incorrect.\n", type);
+                        return kTfLiteError;
+                        }
+                  }
+            }
             return kTfLiteOk;
           }
         }
diff --git a/tensorflow/python/framework/convert_to_constants.py b/tensorflow/python/framework/convert_to_constants.py
index 10541ed8e3..85aa8a277e 100644
--- a/tensorflow/python/framework/convert_to_constants.py
+++ b/tensorflow/python/framework/convert_to_constants.py
@@ -871,11 +871,16 @@ class _SessionConverterData(_ConverterData):
         nodes_to_convert.append(node)
         tensor_names_to_convert.append(tensor_name + ":0")

+    print("Simrit Inside _SessionConverterData")
     if tensor_names_to_convert:
       converted_tensors = session.run(tensor_names_to_convert)
       for node, tensor_value in zip(nodes_to_convert, converted_tensors):
-        self._tensor_data[node.name] = _TensorData(
-            numpy=tensor_value, dtype=node.attr["dtype"].type, index=None)
+        if sys.byteorder == "little":
+            self._tensor_data[node.name] = _TensorData(
+                numpy=tensor_value, dtype=node.attr["dtype"].type, index=None)
+        else:
+            self._tensor_data[node.name] = _TensorData(
+                numpy=np.float32.byteswap(tensor_value), dtype=node.attr["dtype"].type, index=None)


 def disable_lower_using_switch_merge(graph_def):

```
However, I would be more comfortable if this part is handled by SME like yourself so that TF data paths are appropriately covered. If you think it is a reasonable request then I will be happy to assist with testing the changes on BE platform.
Let me know your thoughts. Thanks.

Reply:
Yeah, my main concern is the swizzling of the constants in interpreter_builder. Models (flatbuffer memory references) can be used across multiple interpreters, so even if we can reliably swizzle the weights at that place in the code, it's not a safe thing to be doing generally.

Reply:
Thanks @jdduke - would your team be willing to pursue this further as impact appears much broader than originally thought? We can withdraw the PR and open an issue instead if that helps.

Reply:
@jdduke Can you please take a look on the above comment from @Simrit-Kaur. Thanks!

Reply:
Apologies for the delay, @skribm9 yes if you could file an issue (feel free to assign to me) that would help for tracking purposes. We likely won't get to this in Q4, but we can start thinking about the simplest path to unblock this use-case. Thanks!

Reply:
Thank you @jdduke. I understand that it might take some time to unblock this use-case. I have filed an issue #45009 for tracking purposes.
Thanks!